{
  "url": "https://api.github.com/repos/shap/shap/issues/1644",
  "repository_url": "https://api.github.com/repos/shap/shap",
  "labels_url": "https://api.github.com/repos/shap/shap/issues/1644/labels{/name}",
  "comments_url": "https://api.github.com/repos/shap/shap/issues/1644/comments",
  "events_url": "https://api.github.com/repos/shap/shap/issues/1644/events",
  "html_url": "https://github.com/shap/shap/issues/1644",
  "id": 762544110,
  "node_id": "MDU6SXNzdWU3NjI1NDQxMTA=",
  "number": 1644,
  "title": "LightGBM categoricals and shap_interactions",
  "user": {
    "login": "mayer79",
    "id": 16206095,
    "node_id": "MDQ6VXNlcjE2MjA2MDk1",
    "avatar_url": "https://avatars.githubusercontent.com/u/16206095?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/mayer79",
    "html_url": "https://github.com/mayer79",
    "followers_url": "https://api.github.com/users/mayer79/followers",
    "following_url": "https://api.github.com/users/mayer79/following{/other_user}",
    "gists_url": "https://api.github.com/users/mayer79/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/mayer79/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/mayer79/subscriptions",
    "organizations_url": "https://api.github.com/users/mayer79/orgs",
    "repos_url": "https://api.github.com/users/mayer79/repos",
    "events_url": "https://api.github.com/users/mayer79/events{/privacy}",
    "received_events_url": "https://api.github.com/users/mayer79/received_events",
    "type": "User",
    "site_admin": false
  },
  "labels": [

  ],
  "state": "open",
  "locked": false,
  "assignee": null,
  "assignees": [

  ],
  "milestone": null,
  "comments": 3,
  "created_at": "2020-12-11T16:08:39Z",
  "updated_at": "2020-12-21T14:57:05Z",
  "closed_at": null,
  "author_association": "NONE",
  "active_lock_reason": null,
  "body": "Thanks a lot for your awesome package.\r\n\r\nI keep having troubles with LightGBM's `categorical_feature` support (the data is integer coded, but the algo internally treats them as unordered) and SHAP interactions. \r\n\r\nThe error is most certainly related to this issue on CatBoost: https://github.com/slundberg/shap/issues/1438\r\n\r\nHere is a small example with shap 0.37 and LightGBM 3.1.0\r\n\r\n``` python\r\n\r\nimport numpy as np\r\nimport lightgbm as lgb  # 3.1.0\r\nimport shap  # 0.37.0\r\n\r\n# Create data\r\nn = 1000\r\n\r\nX = np.column_stack((\r\n    np.random.normal(size=n),\r\n    np.random.randint(0, 2, size=n),\r\n    np.random.randint(0, 10, size=n)\r\n))\r\n\r\ny = X.sum(axis=1) + np.random.normal(size=n)\r\n\r\n# Model with categorical features\r\ndtrain = lgb.Dataset(\r\n    X, \r\n    label=y, \r\n    categorical_feature=[1, 2]  # Comment out this line will work\r\n)\r\n\r\nparams = {\r\n    'learning_rate': 0.05,\r\n    'num_leaves': 15,\r\n    'objective': 'mse',\r\n}\r\n \r\n# Fit on best params\r\ngbm = lgb.train(\r\n    params, \r\n    train_set=dtrain, \r\n    num_boost_round=10\r\n)\r\n\r\n# Calculate SHAP\r\nexplainer = shap.TreeExplainer(gbm)\r\nshap_values = explainer.shap_values(X)\r\nshap.summary_plot(shap_values, X, plot_type='bar', color='orange')\r\n\r\n# These lines only work without categorical features\r\nshap_interactions = explainer.shap_interaction_values(X)\r\nshap.summary_plot(shap_interactions, X, plot_type='bar', color='orange')\r\n```\r\n\r\nThis gives\r\n```\r\n---------------------------------------------------------------------------\r\nAttributeError                            Traceback (most recent call last)\r\n<ipython-input-71-87b86078cf72> in <module>\r\n      1 # Calculate N_SHAP decompositions\r\n----> 2 shap_inter = explainer.shap_interaction_values(X_train[:N_SHAP])\r\n\r\n~\\Anaconda3\\lib\\site-packages\\shap\\explainers\\_tree.py in shap_interaction_values(self, X, y, tree_limit)\r\n    470 \r\n    471         if tree_limit < 0 or tree_limit > self.model.values.shape[0]:\r\n--> 472             tree_limit = self.model.values.shape[0]\r\n    473 \r\n    474         # run the core algorithm using the C extension\r\n\r\nAttributeError: 'TreeEnsemble' object has no attribute 'values'\r\n```\r\n\r\n- Everything works without `categorical_features` \r\n- Usual SHAP decompositions work with `categorical_features` as well.\r\n\r\nThe same issue popped up with very differently structured data as well. However, it is not impossible that I keep doing the same stupid stuff over and over ;).",
  "closed_by": null,
  "reactions": {
    "url": "https://api.github.com/repos/shap/shap/issues/1644/reactions",
    "total_count": 2,
    "+1": 2,
    "-1": 0,
    "laugh": 0,
    "hooray": 0,
    "confused": 0,
    "heart": 0,
    "rocket": 0,
    "eyes": 0
  },
  "timeline_url": "https://api.github.com/repos/shap/shap/issues/1644/timeline",
  "performed_via_github_app": null,
  "state_reason": null
}
